### DISCUSSION TITLE
Problem with tween Linear transition

### DISCUSSION THREAD

**Discussion by AcevedoC17** [ramokz/phantom-camera#115] [Question / Help Wanted] [answered]
I have 4 PhantomCams3D, 3 of them transition properly between eachother, but when phantomcam3 transitions to phantomcam 4, the transition rotates along the "long away around" doing an entire 360 instead of just rotating the simpler shorter way like the others. It happens only between PhantomCam3 and PhantomCam4.

PhantomCam3 is on the right, PhantomCam4 is on the left

Any ideas on what causes the issue and/or how to resolve it? The tool has been fantastic and would love for this to work

![image](https://github.com/ramokz/phantom-camera/assets/77506847/f88ada26-eea5-4195-95fc-6fd15b7bfc34)

**Comment by ramokz** [comment, marked as answer]
That's because I've been a bit silly with the way interpolation works for 3D rotation and didn't do it the proper way.

Am refactoring how the tweening and lerping system works in 0.5, so should resolve this, among other, issues.

For now, would suggest either grabbing the master branch, which has this fix but _might_ be a bit unstable, or alternatively inject following bits into `phantom_camera_host.gd`:

_Note:_ The below code will very likely cause errors if you're also using `PhantomCamera2D` in 2D scenes, as this code is 3D specific. But you can surround the various bits with `if _is_3D:` conditionals to solve that.


1. Add a new variable called `_previous_active_pcam_transform` below `_previous_active_pcam_rotation` on line 23.
2. Assign the value as `_previous_active_pcam_transform = camera.get_transform()` below `_previous_active_pcam_rotation` on line 98.
3. Like step 2, do the same below `_previous_active_pcam_rotation` on line 117.
4. On line 144, comment out the entire `camera.set_rotation()` method and inject the below instead:


``` GDScript

var previous_active_pcam_basis = Quaternion(_previous_active_pcam_transform.basis.orthonormalized())
camera.set_quaternion(
	Tween.interpolate_value(
		previous_active_pcam_basis, \
		previous_active_pcam_basis.inverse() * Quaternion(_active_pcam.basis.orthonormalized()),
		tween_duration, \
		_active_pcam.get_tween_duration(), \
		_active_pcam.get_tween_transition(),
		_active_pcam.get_tween_ease(),
	)
)
```

That should mimic the behaviour in the master branch for this property and rotate the camera along the shortest route.

**Reply by AcevedoC17** [reply]
Thank you so much for the detailed reply. Tried it out and it works amazingly!

**Reply by ramokz** [reply]
Glad to hear it!

**Comment by ramokz** [comment]
Code injection solve is written above.

Proper implementation will be in v0.5.
